crux/src/Crux/Goal.hs

```diff
@@ -129,24 +135,121 @@ proveToGoal _ allAsmps p pr =
  where
  showLabPred x = (x^.labeledPredMsg, show (printSymExpr (x^.labeledPred)))
 
+-- | Discharge a tree of proof obligations ('Goals') by using a non-online solver
+--
+-- This function traverses the 'Goals' tree while keeping track of a collection
+-- of assumptions in scope for each goal.  For each proof goal encountered in
+-- the tree, it creates a fresh solver connection using the provided solver
+-- adapter.
+--
+-- This is in contrast to 'proveGoalsOnline', which uses an online solver
+-- connection with scoped assumption frames.  This function allows using a wider
+-- variety of solvers (i.e., ones that don't have support for online solving)
+-- and would in principle enable parallel goal evaluation (though the tree
+-- structure makes that a bit trickier, it isn't too hard).
+--
+-- Note that this function uses the same symbolic backend ('ExprBuilder') as the
+-- symbolic execution phase, which should not be a problem.
+proveGoalsOffline :: forall st sym p asmp ast t fs
+                   . (?outputConfig :: OutputConfig, sym ~ ExprBuilder t st fs)
+                  => WS.SolverAdapter st
+                  -> CruxOptions
+                  -> SimCtxt sym p
+                  -> Maybe (Goals (LPred sym asmp) (LPred sym ast))
+                  -> IO (Maybe (Goals (LPred sym asmp) (LPred sym ast, ProofResult (Either (LPred sym asmp) (LPred sym ast)))))
+proveGoalsOffline _adapter _opts _ctx Nothing = return Nothing
+proveGoalsOffline adapter opts ctx (Just gs0) = do
+  goalNum <- newIORef (0, 0, 0)
+  unless hasUnsatCores $ do
```

Comment:
Probably don't need this warning for offline solving, which doesn't appear to support unsat cores at all.

Reply:
Thanks